src/core/UniswapV4.sol

```diff
@@ -139,72 +149,52 @@ library NotesLib {
         }
     }
 
-    /// This function handles all cases except overflow of the number of tokens being tracked (above
-    /// `_MAX_TOKENS`). In that case we will `Panic` with code 0x32 (array out-of-bounds access).
-    /// This function is largely responsible for maintaining the consistency of the indirection
-    /// array, the heap of `Note` objects, and the transient storage mapping.
-    function insert(Note[] memory a, IERC20 token) internal returns (NotePtr x) {
+    function get(Note[] memory a, IERC20 token, uint256 hashMul, uint256 hashMod) internal pure returns (NotePtr x) {
         assembly ("memory-safe") {
             token := and(_ADDRESS_MASK, token)
-            x := tload(token)
-            switch x
-            case 0 {
-                // We've never seen this token before; we must add it to the mapping and initialize
-
-                // Increment the length of `a`
-                let i := add(0x01, mload(a))
-                mstore(a, i)
-                // `i` is now 1-indexed; we do not require bounds-checking on `i` because it is
-                // implicitly bounded above by `noteAllocations`, which is bounds-checked
-
-                // Find the first free `Note` object (in memory after `a`)
-                let indirectArrayEnd := add(add(0x20, shl(0x05, _MAX_TOKENS)), a)
-                let noteAllocations := mload(indirectArrayEnd)
-                if eq(noteAllocations, _MAX_TOKENS) {
-                    mstore(0x00, 0x4e487b71) // selector for `Panic(uint256)`
-                    mstore(0x20, 0x32) // array out of bounds
-                    revert(0x1c, 0x24)
-                }
-                x := add(indirectArrayEnd, add(0x20, shl(0x06, noteAllocations)))
-                // Allocate it
-                mstore(indirectArrayEnd, add(0x01, noteAllocations))
+            x :=
+                add(a, add(add(0x20, shl(0x05, _MAX_TOKENS)), mod(mulmod(token, hashMul, hashMod), shl(0x06, _MAX_TOKENS))))
+            let oldToken := mload(x)
+            if mul(oldToken, xor(oldToken, token)) { // TODO(dekz): check me on this?
```

Comment:
Dangit. UniV4's use of `address(0)` to represent ETH screwed me here. I've gotta rethink this or add logic to substitute `address(0)` for `0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE` on all the pool manager calls :weary: